app/types/type-structures.ts

```diff
@@ -0,0 +1,317 @@
+export interface Type {
+    subtype(that: Type): boolean;
+    rsubtype?(that: Type): boolean;
+    feed(types: Map<string, Type>): Type;
+    validate(): void;
+    kind: string;
+}
+
+export class TypeScope {
+    constructor(public definitions: Map<string, Type>) {
+
+    }
+
+    feed(scope: TypeScope): TypeScope {
+        // TODO: make faster?
+        const types = new Map([...this.definitions.entries()].concat([...scope.definitions.entries()]));
+
+        for (const [key, type] of this.definitions.entries()) {
+            this.definitions.set(key, type.feed(types));
+        }
+        return this;
+    }
+}
+
+export class TypeVariable implements Type {
+    kind = "Variable";
+
+    constructor(public type?: Type, public lookupName?: string, public matchName?: string) {
+
+    }
+    subtype(that: Type): boolean {
+        if (!this.type) {
+            this.type = that;
+            return true;
+        }
+        if (this.type.subtype(that)) {
+            return true;
+        }
+        if (that.subtype(this.type)) {
+            this.type = that;
+            return true;
+        }
+        return false;
+    }
+
+    rsubtype(that: Type) {
```

Comment:
Personally, I would have subtypeing return a new type so that you still have access to the old one if needed and it seems like it wouldn't be too much of a performance hit.

I also don't like combining read and write methods.